### ISSUE TITLE
Duplicate/overlap of IP on pods across different nodes

### ISSUE THREAD

**Issue by ankycooper** [weaveworks/weave#3793]
What happened:
Duplicate/overlap of IP on pods across different nodes

NAME READY STATUS RESTARTS AGE IP NODE NOMINATED NODE READINESS GATES pod/http-app-5947d79757-4gwd6 1/1 Running 0 20s 10.4.192.2 w2.k8s.local <none> <none> pod/http-app-5947d79757-5ldqx 1/1 Running 0 20s 10.4.192.5 w2.k8s.local <none> <none> pod/http-app-5947d79757-b7lwr 1/1 Running 0 4m11s 10.4.192.1 w1.k8s.local <none> <none> pod/http-app-5947d79757-clswr 1/1 Running 0 20s 10.4.192.4 w2.k8s.local <none> <none> pod/http-app-5947d79757-g79vk 1/1 Running 0 20s 10.4.192.3 w1.k8s.local <none> <none> pod/http-app-5947d79757-g7z2t 1/1 Running 0 20s 10.4.192.4 w1.k8s.local <none> <none> pod/http-app-5947d79757-l6qpb 1/1 Running 0 4m11s 10.4.192.2 w1.k8s.local <none> <none> pod/http-app-5947d79757-ldh8b 1/1 Running 0 4m11s 10.4.192.1 w2.k8s.local <none> <none> pod/http-app-5947d79757-lwmqq 1/1 Running 0 20s 10.4.192.3 w2.k8s.local <none> <none> pod/http-app-5947d79757-wsqd2 1/1 Running 0 20s 10.4.192.5 w1.k8s.local <none> <none>
What you expected to happen:
Pods on different nodes should not have overlapping IPs
How to reproduce it (as minimally and precisely as possible):
Deploy the cluster on ubuntu 18.04
`# On Master
export API_ADDR="10.20.20.9"
export DNS_DOMAIN="k8s.local"
export POD_NET="10.4.0.0/16"
export SRV_NET="10.5.0.0/16"

kubeadm init --pod-network-cidr ${POD_NET} --service-cidr ${SRV_NET} --service-dns-domain "${DNS_DOMAIN}" --apiserver-advertise-address ${API_ADDR}
kubectl apply -f "https://cloud.weave.works/k8s/net?k8s-version=$(kubectl version | base64 | tr -d '\n')&env.IPALLOC_RANGE=${POD_NET}"

#On Workers (2 Workers)
kubeadm join 10.20.20.9:6443 --token --discovery-token-ca-cert-hash sha256:
`

Once the cluster is ready create a test deployment
kind: Deployment apiVersion: apps/v1 metadata: name: http-app namespace: test labels: name: http-app spec: replicas: 3 selector: matchLabels: name: http-app template: metadata: labels: name: http-app spec: terminationGracePeriodSeconds: 60 containers: - name: http-app image: katacoda/docker-http-server:latest

Anything else we need to know?:

Environment: 3VMs ubuntu 18.04 (1 master + 2 Workers)

Kubernetes version (use kubectl version): Client Version: v1.18.1 | Server Version: v1.18.1
Cloud provider or hardware configuration: 4cpu +4 Gb ram + 100G HDD
OS (e.g: cat /etc/os-release): Ubuntu 18.04.4 LTS (Bionic Beaver)
Kernel (e.g. uname -a): 4.15.0-96-generic
Install tools: kubeadm
Network plugin and version (if this is a network-related bug): Weave net
Others: Calico works fine

## Logs:
Running Weave Pods
```
NAME              READY   STATUS    RESTARTS   AGE   IP           NODE                               NOMINATED NODE   READINESS GATES
weave-net-dmt77   2/2     Running   1          11h   10.20.20.9   master-node.k8s.local   <none>           <none>
weave-net-krb7c   2/2     Running   1          11h   10.20.20.8   w1.k8s.local            <none>           <none>
weave-net-ppvmn   2/2     Running   1          11h   10.20.20.7   w2.k8s.local            <none>           <none>
```
Logs of Pod on master 
`INFO: 2020/04/11 09:38:15.450678 ->[10.20.20.8:47541] connection accepted
INFO: 2020/04/11 09:38:15.451216 ->[10.20.20.8:47541|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection ready; using protocol version 2
INFO: 2020/04/11 09:38:15.451326 overlay_switch ->[3a:8a:4e:83:b9:2a(w2.k8s.local)] using fastdp
INFO: 2020/04/11 09:38:15.451470 ->[10.20.20.8:47541|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection shutting down due to error: Multiple connections to 3a:8a:4e:83:b9:2a(w2.k8s.local) added to c6:94:4f:41:b9:42(master-node.k8s.local)
INFO: 2020/04/11 09:47:08.858431 ->[10.20.20.8:49939] connection accepted
INFO: 2020/04/11 09:47:08.859080 ->[10.20.20.8:49939|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection ready; using protocol version 2
INFO: 2020/04/11 09:47:08.859195 overlay_switch ->[3a:8a:4e:83:b9:2a(w2.k8s.local)] using fastdp
INFO: 2020/04/11 09:47:08.859289 ->[10.20.20.8:49939|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection shutting down due to error: Multiple connections to 3a:8a:4e:83:b9:2a(w2.k8s.local) added to c6:94:4f:41:b9:42(master-node.k8s.local)
INFO: 2020/04/11 09:51:11.836319 ->[10.20.20.8:47963] connection accepted
INFO: 2020/04/11 09:51:11.837113 ->[10.20.20.8:47963|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection ready; using protocol version 2
INFO: 2020/04/11 09:51:11.837222 overlay_switch ->[3a:8a:4e:83:b9:2a(w2.k8s.local)] using fastdp
INFO: 2020/04/11 09:51:11.837293 ->[10.20.20.8:47963|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection shutting down due to error: Multiple connections to 3a:8a:4e:83:b9:2a(w2.k8s.local) added to c6:94:4f:41:b9:42(master-node.k8s.local)
INFO: 2020/04/11 09:57:00.570089 ->[10.20.20.8:58447] connection accepted
INFO: 2020/04/11 09:57:00.570690 ->[10.20.20.8:58447|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection ready; using protocol version 2
INFO: 2020/04/11 09:57:00.570842 overlay_switch ->[3a:8a:4e:83:b9:2a(w2.k8s.local)] using fastdp
INFO: 2020/04/11 09:57:00.570872 ->[10.20.20.8:58447|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection shutting down due to error: Multiple connections to 3a:8a:4e:83:b9:2a(w2.k8s.local) added to c6:94:4f:41:b9:42(master-node.k8s.local)
INFO: 2020/04/11 10:04:19.595463 ->[10.20.20.8:53187] connection accepted
INFO: 2020/04/11 10:04:19.596120 ->[10.20.20.8:53187|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection ready; using protocol version 2
INFO: 2020/04/11 10:04:19.596231 overlay_switch ->[3a:8a:4e:83:b9:2a(w2.k8s.local)] using fastdp
INFO: 2020/04/11 10:04:19.596347 ->[10.20.20.8:53187|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection shutting down due to error: Multiple connections to 3a:8a:4e:83:b9:2a(w2.k8s.local) added to c6:94:4f:41:b9:42(master-node.k8s.local)
INFO: 2020/04/11 10:12:02.827258 ->[10.20.20.8:60041] connection accepted
INFO: 2020/04/11 10:12:02.827909 ->[10.20.20.8:60041|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection ready; using protocol version 2
INFO: 2020/04/11 10:12:02.827970 overlay_switch ->[3a:8a:4e:83:b9:2a(w2.k8s.local)] using fastdp
INFO: 2020/04/11 10:12:02.827993 ->[10.20.20.8:60041|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection shutting down due to error: Multiple connections to 3a:8a:4e:83:b9:2a(w2.k8s.local) added to c6:94:4f:41:b9:42(master-node.k8s.local)
INFO: 2020/04/11 10:20:15.465463 ->[10.20.20.8:51247] connection accepted
INFO: 2020/04/11 10:20:15.465951 ->[10.20.20.8:51247|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection ready; using protocol version 2
INFO: 2020/04/11 10:20:15.465999 overlay_switch ->[3a:8a:4e:83:b9:2a(w2.k8s.local)] using fastdp
INFO: 2020/04/11 10:20:15.466039 ->[10.20.20.8:51247|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection shutting down due to error: Multiple connections to 3a:8a:4e:83:b9:2a(w2.k8s.local) added to c6:94:4f:41:b9:42(master-node.k8s.local)
INFO: 2020/04/11 10:28:48.032877 Discovered remote MAC 06:1a:e7:27:5d:64 at 3a:8a:4e:83:b9:2a(w2.k8s.local)
INFO: 2020/04/11 10:29:15.092271 ->[10.20.20.8:42009] connection accepted
INFO: 2020/04/11 10:29:15.092789 ->[10.20.20.8:42009|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection ready; using protocol version 2
INFO: 2020/04/11 10:29:15.092932 overlay_switch ->[3a:8a:4e:83:b9:2a(w2.k8s.local)] using fastdp
INFO: 2020/04/11 10:29:15.093036 ->[10.20.20.8:42009|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection shutting down due to error: Multiple connections to 3a:8a:4e:83:b9:2a(w2.k8s.local) added to c6:94:4f:41:b9:42(master-node.k8s.local)
INFO: 2020/04/11 10:30:59.105836 Discovered remote MAC 12:ff:c5:d2:c5:22 at 3a:8a:4e:83:b9:2a(w2.k8s.local)
INFO: 2020/04/11 10:32:40.976961 ->[10.20.20.8:38145] connection accepted
INFO: 2020/04/11 10:32:40.977755 ->[10.20.20.8:38145|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection ready; using protocol version 2
INFO: 2020/04/11 10:32:40.977898 overlay_switch ->[3a:8a:4e:83:b9:2a(w2.k8s.local)] using fastdp
INFO: 2020/04/11 10:32:40.977921 ->[10.20.20.8:38145|3a:8a:4e:83:b9:2a(w2.k8s.local)]: connection shutting down due to error: Multiple connections to 3a:8a:4e:83:b9:2a(w2.k8s.local) added to c6:94:4f:41:b9:42(master-node.k8s.local)
INFO: 2020/04/11 10:33:10.178578 Discovered remote MAC 3a:8a:4e:83:b9:2a at 3a:8a:4e:83:b9:2a(w2.k8s.local)
`

## Network:

$ ip route
master
`default via 10.20.20.1 dev ens160 proto static
10.4.0.0/16 dev weave proto kernel scope link src 10.4.0.1
10.20.20.0/24 dev ens160 proto kernel scope link src 10.20.20.9
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 linkdown
`
w1
`
default via 10.20.20.1 dev ens160 proto static
10.4.0.0/16 dev weave proto kernel scope link src 10.4.192.0
10.20.20.0/24 dev ens160 proto kernel scope link src 10.20.20.8
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 linkdown
`

w2
`
default via 10.20.20.1 dev ens160 proto static
10.4.0.0/16 dev weave proto kernel scope link src 10.4.192.0
10.20.20.0/24 dev ens160 proto kernel scope link src 10.20.20.7
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 linkdown
`

$ ip -4 -o addr

master
`1: lo    inet 127.0.0.1/8 scope host lo\       valid_lft forever preferred_lft forever
2: ens160    inet 10.20.20.9/24 brd 10.20.20.255 scope global ens160\       valid_lft forever preferred_lft forever
3: docker0    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0\       valid_lft forever preferred_lft forever
6: weave    inet 10.4.0.1/16 brd 10.4.255.255 scope global weave\       valid_lft forever preferred_lft forever
`
w1
`1: lo    inet 127.0.0.1/8 scope host lo\       valid_lft forever preferred_lft forever
2: ens160    inet 10.20.20.8/24 brd 10.20.20.255 scope global ens160\       valid_lft forever preferred_lft forever
3: docker0    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0\       valid_lft forever preferred_lft forever
6: weave    inet 10.4.192.0/16 brd 10.4.255.255 scope global weave\       valid_lft forever preferred_lft forever
`
w2
`
1: lo    inet 127.0.0.1/8 scope host lo\       valid_lft forever preferred_lft forever
2: ens160    inet 10.20.20.7/24 brd 10.20.20.255 scope global ens160\       valid_lft forever preferred_lft forever
3: docker0    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0\       valid_lft forever preferred_lft forever
6: weave    inet 10.4.192.0/16 brd 10.4.255.255 scope global weave\       valid_lft forever preferred_lft forever
`

**Comment by ankycooper** [comment]
K8S allocated each node with a different /24 from the /16 CIDR however weave does not comply with this

`# kubectl get nodes master-node.k8s.local -o json |grep "10.4"
                            "v:\"10.4.0.0/24\"": {}
        "podCIDR": "10.4.0.0/24",
            "10.4.0.0/24"
# kubectl get nodes w1.k8s.local -o json |grep "10.4"
                            "v:\"10.4.1.0/24\"": {}
        "podCIDR": "10.4.1.0/24",
            "10.4.1.0/24"
# kubectl get nodes w2.k8s.local -o json |grep "10.4"
                            "v:\"10.4.2.0/24\"": {}
        "podCIDR": "10.4.2.0/24",
            "10.4.2.0/24"
`
Weave issues duplicate/ overlapping IPs to pods on differnet nodes.

**Comment by ankycooper** [comment]
Found the solution!!

I had cloned the VMs for some reason the interface mac did not change for W1 and W2
Even after changing the interface mac and rebuilding the cluster, there was no difference in behaviour and weave interface on W1 and W2 had the same mac.


checked /etc/machine-id on all nodes and it was same. 

deleted the machine-id files 
`
truncate -s 0 /etc/machine-id
rm /var/lib/dbus/machine-id
init 6
`
rebooted all nodes & rebuild the cluster.

Boom its working. Cloning did not do the clean-up.
